src/chigame/users/models.py

```diff
@@ -93,24 +93,68 @@ class GroupInvitation(models.Model):
 
 
 class NotificationQuerySet(models.QuerySet):
-    def filter_by_actor(self, actor, **kwargs):
+    def filter_by_actor(self, actor, include_deleted=False, **kwargs):
         try:
             actor_content_type = ContentType.objects.get(model=actor._meta.model_name)
             actor_object_id = actor.pk
-            return self.filter(actor_content_type=actor_content_type, actor_object_id=actor_object_id, **kwargs)
+            queryset = self.filter(actor_content_type=actor_content_type, actor_object_id=actor_object_id, **kwargs)
+            if not include_deleted:
+                queryset = queryset.is_not_deleted()
+            return queryset
 
         except ContentType.DoesNotExist:
             raise ValueError(f"The model {actor.label} is not registered in content type")
 
-    def get_by_actor(self, actor, **kwargs):
+    def get_by_actor(self, actor, include_deleted=False, **kwargs):
         try:
             actor_content_type = ContentType.objects.get(model=actor._meta.model_name)
             actor_object_id = actor.pk
-            return self.get(actor_content_type=actor_content_type, actor_object_id=actor_object_id, **kwargs)
+            queryset = self.get(actor_content_type=actor_content_type, actor_object_id=actor_object_id, **kwargs)
+            if not include_deleted:
+                queryset = queryset.is_not_deleted()
+            return queryset
 
         except ContentType.DoesNotExist:
             raise ValueError(f"The model {actor.label} is not registered in content type")
 
+    def filter_by_receiver(self, receiver, include_deleted=False):
+        queryset = self.filter(receiver=receiver)
+        if not include_deleted:
+            queryset = queryset.is_not_deleted()
+        return queryset
+
+    def filter_by_type(self, type, include_deleted=False):
+        if type not in [type[0] for type in Notification.NOTIFICATION_TYPES]:
+            raise ValueError(f"{type} is not a valid type")
+        queryset = self.filter(type=type)
+        if not include_deleted:
+            queryset = queryset.is_not_deleted()
+        return queryset
+
+    def mark_all_unread(self):
+        self.update(read=False)
+
+    def mark_all_read(self):
+        self.update(read=True)
+
+    def mark_all_deleted(self):
+        self.update(visible=False)
+
+    def restore_all_deleted(self):
+        self.update(visible=True)
+
+    def is_read(self):
+        return self.filter(read=True)
```

Comment:
My only question is when is_read, is_unread, is_deleted, and is_not_deleted, how does the system know to only mark a specific notification as read, unread, deleted or not deleted? Does it need an additional input?

Reply:
these methods are called on querysets so for example for is_read, the queryset will have Notification where read is True

Reply:
And does the filter individualize it so only one is affected, and not all?

Reply:
All Notification that fit this criteria are included in the queryset.

Reply:
I see, thank you for the clarification. the helper methods look properly implemented in the Notification model.